src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -240,149 +238,331 @@ private static ImmutableArray<TextChangeRange> Merge(IReadOnlyList<ImmutableArra
+            {
+                if (oldChange.Span.Length == 0 && oldChange.NewLength == 0)
+                {
+                    // old change does not insert or delete any characters, so it can be dropped to no effect.
+                    if (tryGetNextOldChange()) continue;
+                    else break;
+                }
+                else if (newChange.SpanLength == 0 && newChange.NewLength == 0)
+                {
+                    // new change does not insert or delete any characters, so it can be dropped to no effect.
+                    if (tryGetNextNewChange()) continue;
+                    else break;
+                }
+                else if (newChange.SpanEnd < oldChange.Span.Start + oldDelta)
+                {
+                    // new change is entirely before old change, so just take the new change
+                    //                old[--------]
+                    // new[--------]
+                    addAdjustedNewChange(builder, oldDelta, newChange);
+                    if (tryGetNextNewChange()) continue;
+                    else break;
+                }
+                else if (newChange.SpanStart > oldChange.NewEnd + oldDelta)
+                {
+                    // new change is entirely after old change, so just take the old change
+                    // old[--------]
+                    //                new[--------]
+                    addAndAdjustOldDelta(builder, ref oldDelta, oldChange);
+                    if (tryGetNextOldChange()) continue;
+                    else break;
+                }
+                else if (newChange.SpanStart < oldChange.Span.Start + oldDelta)
+                {
+                    // new change starts before old change, but overlaps
+                    // note: 'd' represents a deleted character, 'a' represents a character inserted by an old change, and 'b' represents a character inserted by a new change.
+                    //
+                    //    old|dddddd|
+                    //       |aaaaaa|
+                    // ---------------
+                    // new|dddddd|
+                    //    |bbbbbb|
+
+                    // align the new change and old change start by consuming the part of the new deletion before the old change
+                    // (this only deletes characters of the original text)
+                    //
+                    //    old|dddddd|
+                    //       |aaaaaa|
+                    // ---------------
+                    //    new|ddd|
+                    //       |bbbbbb|
+                    var newChangeLeadingDeletion = oldChange.Span.Start + oldDelta - newChange.SpanStart;
+                    addAdjustedNewChange(builder, oldDelta, new UnadjustedNewChange(newChange.SpanStart, newChangeLeadingDeletion, newLength: 0));
+                    newChange = new UnadjustedNewChange(oldChange.Span.Start + oldDelta, newChange.SpanLength - newChangeLeadingDeletion, newChange.NewLength);
```

Comment:
Maybe I'm missing something but I believe this `if` block only works when `newChangeLedaingDeletion` is essentially non-zero here. I can't see how that guarantee is met though. What am i missing?

Reply:
This form of overlap can only occur when the new change has deletions. If the new change did not have deletions, and started before the old change, then the new change's insertions would simply push the old changes to the right rather than overlapping with them:

```
01234
  dd  // old change
  aa
-----
01aa4
 bbb  // new change with no deletion
-----
0bbb1aa4
```

(also, after chat with @jcouv, I think ths illustration would be a lot more clear if we shifted the old change to the left rather than shifting the new change to the right..)